### PR TITLE
Fixed #23916 -- Allowed makemigrations to handle related model name case changes.

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket](https://code.djangoproject.com/ticket/23916)

I believe this is the correct fix.

The reason continuous migrations were being generated was the comparison of the before and after deconstructed `ForeignKey`s in `MigrationAutodetector.generate_altered_fields`:

```
if old_field_dec != new_field_dec:
```

The model name in the migration history is based on the migration that added the model, but nothing ever changes this casing since the autodetector only uses the "model name" (lowercased class name) throughout. `ForeignObject.deconstruct()` seems to have been a place where this was missed.

### CODE REVIEW

**Comment by adamchainz** [comment]
Oh yes ofc

I added the below to the autodetector tests but it already passed before the patch. I think it's because the error wasn't in the autodetector but in the difference between model states / deconstructed fields loaded from migration history and actual models.

```python
    def test_model_case_change(self):
        AuthoR = ModelState('testapp', 'AuthoR', [("id", models.AutoField(primary_key=True))])
        changes = self.get_changes([self.author_empty, self.book], [AuthoR, self.book])
        self.assertNumberMigrations(changes, 'app', 0)
```

That said, this patch does fix the problem on $clientproject that had the same bug as the ticket.

I will come back to this tomorrow to try add a unit test that does reproduce the scenario.

**Comment by adamchainz** [comment]
I'm finding it too hard to make a replicating test. Tried this as well but it passes on master.

```python
    def test_rename_model_case(self):
        """
        Tests that a model changing case doesn't lead to any autodetected
        operations.
        """
        author1 = ModelState('testapp', 'author', [("id", models.AutoField(primary_key=True))])
        book1 = ModelState("otherapp", "Book", [
            ("id", models.AutoField(primary_key=True)),
            ("author", models.ForeignKey("testapp.Author", models.CASCADE)),
        ])
        author2 = ModelState('testapp', 'Author', [("id", models.AutoField(primary_key=True))])
        book2 = ModelState("otherapp", "Book", [
            ("id", models.AutoField(primary_key=True)),
            ("author", models.ForeignKey("testapp.Author", models.CASCADE)),
        ])
        changes = self.get_changes([author1, book1], [author2, book2])
        self.assertNumberMigrations(changes, 'app', 0)
```

🤷‍♂

**Comment by felixxm** [comment]
> I'm finding it too hard to make a replicating test. Tried this as well but it passes on master.

I will play with this.

File: django/db/models/fields/related.py [hunk 1]
```diff
@@ -1561,7 +1566,9 @@ def contribute_to_class(self, cls, name, **kwargs):
         # automatically. The funky name reduces the chance of an accidental
         # clash.
         if self.remote_field.symmetrical and (
-                self.remote_field.model == "self" or self.remote_field.model == cls._meta.object_name):
+            self.remote_field.model == RECURSIVE_RELATIONSHIP_CONSTANT or
+            self.remote_field.model == cls._meta.object_name
+        ):
```
Comment (felixxm) [inline django/db/models/fields/related.py]: This change is unrelated, I moved it to the #12328 and changed all hard-coded values to the constant.

**Comment by felixxm** [comment]
@MarkusH Can you take a look? At first glance, I would expect to get two operations: `RenameModel` and `AlterField` :thinking: , but with this PR we ignore case changes.

File: django/db/models/fields/related.py (hunk 1 above)
Comment (felixxm) [reply django/db/models/fields/related.py]: Merged in 1e0dcd6c8bfa4519c21014c73eb510620dd1a000.

Comment (adamchainz) [reply django/db/models/fields/related.py]: 👍  rebased

File: tests/field_deconstruction/tests.py [hunk 2]
```diff
@@ -196,53 +196,53 @@ def test_float_field(self):
     def test_foreign_key(self):
         # Test basic pointing
         from django.contrib.auth.models import Permission
-        field = models.ForeignKey("auth.Permission", models.CASCADE)
+        field = models.ForeignKey("auth.permission", models.CASCADE)
```
Comment (MarkusH) [inline tests/field_deconstruction/tests.py]: Why this change?

Comment (adamchainz) [reply tests/field_deconstruction/tests.py]: Unnecessary, undone, thanks for spotting.

**Comment by adamchainz** [comment]
Verified it won't trigger the generation of new migrations with a test project: https://github.com/adamchainz/django-23916 .

File: django/db/models/fields/related.py [hunk 3]
```diff
@@ -572,11 +572,16 @@ def deconstruct(self):
...
+            if "." in self.remote_field.model:
+                app_label, _dot, rest = self.remote_field.model.partition(".")
+                kwargs['to'] = "%s.%s" % (app_label, rest.lower())
+            else:
+                kwargs['to'] = self.remote_field.model.lower()
         else:
             kwargs['to'] = "%s.%s" % (
                 self.remote_field.model._meta.app_label,
-                self.remote_field.model._meta.object_name,
+                self.remote_field.model._meta.model_name,
```
Comment (charettes) [inline django/db/models/fields/related.py]: This could be replaced by `self.remote_field.model._meta.label_lower`

https://github.com/django/django/blob/c1b24718e05ea474955777d7bc4d9d5634560cd5/django/db/models/options.py#L136-L138

Comment (adamchainz) [reply django/db/models/fields/related.py]: Done!

**Comment by charettes** [comment]
Is the only thing blocking the merge of this change a valid regression test? If it's the case I could likely reproduce if there was a simplified test case.

**Comment by felixxm** [comment]
@charettes Yes, a regression test is missing.

**Comment by adamchainz** [comment]
Took another pass. Found a way to fix up my above test to fail on master and pass on this PR. Mostly it was my call to `assertNumberMigrations` using the wrong app label(s) 🤦‍♂️

File: django/db/models/fields/related.py [hunk 4]
```diff
@@ -583,12 +583,14 @@ def deconstruct(self):
             kwargs['parent_link'] = self.remote_field.parent_link
         # Work out string form of "to"
         if isinstance(self.remote_field.model, str):
-            kwargs['to'] = self.remote_field.model
+            # Force model name to lower case
+            if "." in self.remote_field.model:
+                app_label, _dot, rest = self.remote_field.model.partition(".")
+                kwargs['to'] = "%s.%s" % (app_label, rest.lower())
+            else:
+                kwargs['to'] = self.remote_field.model.lower()
```
Comment (felixxm) [inline django/db/models/fields/related.py]: I think we can simplify this and use:
```python
kwargs['to'] = self.remote_field.model.lower()
```
like we already do in [`Operation`](https://github.com/django/django/blob/895f28f9cbed817c00ab68770433170d83132d90/django/db/migrations/operations/base.py#L128-L134).

File: tests/migrations/test_autodetector.py [hunk 5]
```diff
@@ -1032,6 +1032,29 @@ def test_rename_model(self):
...
+            ("id", models.AutoField(primary_key=True)),
+            ("author", models.ForeignKey("testapp.Author", models.CASCADE)),
+        ])
+        changes = self.get_changes(
+            [author1, book1],
+            [author2, book2],
+            questioner=MigrationQuestioner({'ask_rename_model': True}),
+        )
+        self.assertNumberMigrations(changes, 'testapp', 0)
+        self.assertNumberMigrations(changes, 'otherapp', 0)
```
Comment (felixxm) [inline tests/migrations/test_autodetector.py]: We can reuse existing states.

**Comment by adamchainz** [comment]
Great, LGTM
